syntheseus/search/algorithms/best_first/base.py

```diff
@@ -2,16 +2,83 @@
 
 import abc
 import heapq
+import itertools
 import logging
 import math
-from typing import Generic, Sequence
+from dataclasses import dataclass, field
+from typing import Any, Generic, Sequence
 
 from syntheseus.search.algorithms.base import GraphType, SearchAlgorithm
 from syntheseus.search.graph.node import BaseGraphNode
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass(order=True)
+class PriorityQueueItem:
+    priority: float
+    tie_breaker: int
+    item: Any = field(compare=False)
+
+
+class PriorityQueue:
+    """
+
+    Simple priority queue implementation which supports removing elements
+    and changing their priority.
+
+    Implementation based on heapq and guidance from:
+    https://docs.python.org/3/library/heapq.html#priority-queue-implementation-notes
+
+    In particular, this means items are never really removed from the queue:
+    they are just marked as invalid and set to `None`. This is why the queue
+    length is not necessarily the same as the number of unique items in the
+    queue. It also assumes that items are unique and hashable.
+    """
+
+    def __init__(self):
+        self._queue = []
+        self._entry_finder = {}
+        self._counter = itertools.count()  # to break ties in priority
+
+    def remove_item(self, item):
+        """Removes and item if it is present."""
+        if item in self._entry_finder:
+            entry = self._entry_finder.pop(item)
+            entry.item = None
+
+    def push_item(self, item, priority):
+        """
+        Pushes an item with a given priority.
+
+        If the item is already present, it is removed and re-inserted.
+        """
+        if item in self._entry_finder:
+            self.remove_item(item)
+        entry = PriorityQueueItem(priority, next(self._counter), item)
+        self._entry_finder[item] = entry
+        heapq.heappush(self._queue, entry)
+
+    def pop_item(self):
+        """Removes an item with the lowest priority and returns it."""
+        while self._queue:
+            entry = heapq.heappop(self._queue)
+            if self._entry_finder.get(entry.item) == entry:
+                # ^^ ensures 1) item in entry finder and
+                # 2) item in entry finder has the same priority as the popped item
+                # (if `None` is added to the queue there may be multiple `None` items)
+                del self._entry_finder[entry.item]
+                return entry
+        raise KeyError("pop from an empty priority queue")
```

Comment:
nit: [`heapq` uses `IndexError` in this case](https://docs.python.org/3/library/heapq.html#heapq.heappop), which may be a slightly better fit